src/System.Management.Automation/namespaces/FileSystemContentStream.cs

```diff
@@ -314,8 +273,29 @@ internal class FileSystemContentReaderWriter : IContentReader, IContentWriter
             bool isRawStream)
             : this(path, streamName, mode, access, share, encoding, false, waitForChanges, provider, isRawStream)
         {
-            _delimiter = delimiter;
-            _usingDelimiter = true;
+            // If the delimiter is default ('\n') we'll use ReadLine() method.
+            // Otherwise allocate temporary structures for ReadDelimited() method.
+            if (!(delimiter.Length == 1 && delimiter[0] == DefaultDelimiter))
+            {
+                _delimiter = delimiter;
+                _usingDelimiter = true;
+
+                // We expect that we are parsing files where line lengths can be relatively long.
+                const int DefaultLineLength = 256;
+                _currentLineContent = new StringBuilder(DefaultLineLength);
+
+                // For Boyer-Moore string search algorithm.
+                // Populate the offset lookups.
+                // These will tell us the maximum number of characters
+                // we can read to generate another possible match.
+                // If we read more characters than this, we risk consuming
+                // more of the stream than we need.
+                _offsetDictionary = new Dictionary<char, int>(_delimiter.Length);
```

Comment:
This looks a little bit dangerous.  If _usingDelimiter is set to true anywhere else, _currentLineContent, _offsetDictionary will be null during later reads.  I feel we should at least have an assert where these are used to ensure they are not null and as a warning if the code somehow gets changed.

Reply:
If somebody make the wrong change our tests will be crashed with null reference exception and we'll never merge the code. I don't know where we could set the assert and how it can help more then the null reference exception.